### PR TITLE
Added basic HL7 functionality to RIS

### PR DESCRIPTION
**Pull request by doc-E-brown**
Basic functionality for sending / receiving HL7 messages from the ClearCanvas RIS server

### CODE REVIEW

File: Ris/Application/Services/HL7/HL7Interface.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+            // Write the expected start characters
+            stream.Write(INTRO_CHARS, 0, INTRO_CHARS.Length);
+
+            stream.Write(data, 0, data.Length);
+
+            // Write the expected end characters
+            stream.Write(EXIT_CHARS, 0, EXIT_CHARS.Length);
+
+            Platform.Log(LogLevel.Info, String.Format("Sent {0}", message));
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7Interface.cs]: Worth including the INTRO_CHARS and EXIT_CHARS in the logs.

File: Ris/Application/Services/HL7/HL7Interface.cs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+            data = new Byte[256];
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7Interface.cs]: Is response guaranteed to be less than 256Bytes?

File: Ris/Application/Services/HL7/HL7Interface.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                var totalBytesReceivedFromClient = 0;
+                int bytesReceived;
+
+                //keeping reading while data is available
+                while ((bytesReceived = stream.Read(receivedBuff, 0, receivedBuff.Length)) > 0)
+                {
+                    if (stream.CanWrite)
+                    {
+                        //echo the received data back to the client 
+                        stream.Write(receivedBuff, 0, bytesReceived);
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7Interface.cs]: The response back does not adhere to the HL7 standard format.
I.e. INTRO_CHARS and EXIT_CHARS. 
Also, if this response is an ACK, it should follow a format similar to the following as an example:

MSH|^~\&|LABADT|DH|EPICADT|DH|201301011228||ACK^A01^ACK |HL7ACK00001|P|2.3
MSA|AA|HL7MSG00001

File: Ris/Application/Services/HL7/HL7Interface.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+                    processingThread.Start(connection);
+                }
+            }
+            catch (Exception ex)
+            {
+                Platform.Log(LogLevel.Error, "Unhandled exception: " + ex.Message);
+            }
+            finally
+            {
+                tcpListener.Stop();
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7Interface.cs]: Worth calling the GC through .Dispose() here?

File: Ris/Application/Services/HL7/HL7Interface.cs (hunk 1 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7Interface.cs]: No I don't think so, these are META characters, used by the machines only and would just take up space in the logs for no value.

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        //private const string PATIENT_LOCATION = "Room";
+
+        // Message Types
+        private const string MSG_TYPE_PROCEDURE_UPDATE = "ORM^O01";
+
+        //Processing IDs
+        private const string ID_DEVELOPMENT = "D";
+
+        // HL7 Version: NOTE: No claims are made regarding adherence to the version
+        private const string HL7_VERSION = "2.3";
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7MessageBuilder.cs]: I thought we were conforming to 2.3.1

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+            return String.Format("{0}\r{1}\r{2}\r{3}\r{4}", msh, pid, pv1, orc, obr);
+        }
+
+        public string getMSHSegment(string patient_account, string message_type, string datetime)
+        {
+            //example: MSH|^~\&|OpenEMR|ChestClinic|||20191004105058||ORM^O01|9|D|2.3
+            return String.Format("MSH|{0}|{1}|{2}|||{3}||{4}|{5}|{6}|{7}",
+                ENCODING_CHARACTERS, SENDING_APPLICATION, this.sending_facility,
+                datetime, message_type, patient_account, ID_DEVELOPMENT, HL7_VERSION
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7MessageBuilder.cs]: datetime is not listed in the Visage 7 spec for the MSH segment.

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                );
+        }
+
+        public string getPIDSegment(string patient_account, string patient_id, string first_name, string last_name, string dob, string sex)
+        {
+            //Example: PID|4|4|4||DUCK^Fluffy||19690930|F|||^^^^|||||||9|345-67-8899|||
+            return String.Format("PID|{0}|{1}|{2}||{3}^{4}||{5}|{6}|||address||||||||{7}|||",
+                patient_id, patient_id, patient_id, 
+                last_name, first_name, dob, sex, patient_account
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7MessageBuilder.cs]: patient account not listed in the Visage 7 PID segment

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                );
+
+        }
+
+        public string getPV1Segment(string accession_number, string referrer_name, string referrer_surname)
+        {
+            //Example: PV1|1||||||^Song^Peter||||||||||||9|||||||||||||
+            return String.Format("PV1|{0}||||||^{1}^{2}||||||||||||{3}|||||||||||||",
+                accession_number, referrer_surname, referrer_name, accession_number
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7MessageBuilder.cs]: accession_number listed twice.

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            this.sending_facility = SENDING_FACILITY;
+        }
+
+        public HL7MessageBuilder(string sending_facility)
+        {
+            this.sending_facility = sending_facility;
+        }
+
+        // Private methods
```
Comment (Chumbaloo) [inline Ris/Application/Services/HL7/HL7MessageBuilder.cs]: remove comment

File: Ris/Server/Application.cs [hunk 10]
```diff
@@ -40,12 +41,21 @@ public class Application : Shred, IApplicationRoot
...
 		#region IApplicationRoot Members
 
 		void IApplicationRoot.RunApplication(string[] args)
 		{
 			StartUp();
 
+            HL7Client client = new HL7Client("192.168.93.235");
+            HL7MessageBuilder msg_builder = new HL7MessageBuilder();
+
+            var hl7_message = msg_builder.getHL7Msg("1234", "id12", "Fluffy", "DUCK", "19600326", "F", "Acc123", "PS", "Peter", "Song", "SC", "CM", "20191004104900", "20191004104901", "2");
```
Comment (Chumbaloo) [inline Ris/Server/Application.cs]: suggestion: to use the named arguments given that the function is overloaded.

File: Ris/Server/Application.cs [hunk 11]
```diff
@@ -113,6 +123,10 @@ private void StartUp()
 			// kick NHibernate, rather than waiting for it to load on demand
 			PersistentStoreRegistry.GetDefaultStore();
 
+            // Start the HL7 server
+            _hl7server = new HL7Server();
+            _hl7server.start();
+
 			_isStarted = true;
```
Comment (Chumbaloo) [inline Ris/Server/Application.cs]: Worth including cleanup once server goes down?

File: Ris/Application/Services/HL7/HL7Interface.cs (hunk 3 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7Interface.cs]: Yes we do, we need to add the additional information from the RIS database.  To be done soon.

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs (hunk 5 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7MessageBuilder.cs]: Read comment above ;)

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs (hunk 6 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7MessageBuilder.cs]: Not an issue, semantic naming differences

File: Ris/Application/Services/HL7/HL7MessageBuilder.cs (hunk 8 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7MessageBuilder.cs]: yes, as we currently dont have a concept of a visit

File: Ris/Server/Application.cs (hunk 10 above)
Comment (doc-E-brown) [reply Ris/Server/Application.cs]: I agree, but this line will be remove

File: Ris/Application/Services/HL7/HL7Interface.cs (hunk 2 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7Interface.cs]: Replaced with a response string that gets populated by the temporary buffer

File: Ris/Application/Services/HL7/HL7Interface.cs (hunk 4 above)
Comment (doc-E-brown) [reply Ris/Application/Services/HL7/HL7Interface.cs]: There is not Dispose method for the tcpListener it does not implement then IDisposable interface.

File: Ris/Server/Application.cs (hunk 11 above)
Comment (doc-E-brown) [reply Ris/Server/Application.cs]: Not required as the listener is blocking, it will be brought down when the program terminates.

**Pull request merged**